pkg/policy/identity.go

```diff
@@ -63,21 +65,82 @@ func (id NumericIdentity) Uint32() uint32 {
 	return uint32(id)
 }
 
-type RuleContexts map[RuleContext]bool
+// SecurityIDContexts maps a security identity to a L4RuleContexts
+type SecurityIDContexts map[NumericIdentity]L4RuleContexts
 
-// RuleContext represents a L3-dependent L4 rule
+func (sc SecurityIDContexts) DeepCopy() SecurityIDContexts {
+	cpy := make(SecurityIDContexts)
+	for k, v := range sc {
+		cpy[k] = v.DeepCopy()
+	}
+	return cpy
+}
+
+// SecurityIDContexts returns a new L4RuleContexts created.
+func NewSecurityIDContexts() SecurityIDContexts {
+	return SecurityIDContexts(make(map[NumericIdentity]L4RuleContexts))
+}
+
+// L4RuleContexts maps a rule context to a boolean.
+type L4RuleContexts map[L4RuleContext]L7RuleContext
+
+// NewL4RuleContexts returns a new L4RuleContexts created.
+func NewL4RuleContexts() L4RuleContexts {
+	return L4RuleContexts(make(map[L4RuleContext]L7RuleContext))
+}
+
+func (rc L4RuleContexts) DeepCopy() L4RuleContexts {
+	cpy := make(L4RuleContexts)
+	for k, v := range rc {
+		cpy[k] = v
+	}
+	return cpy
+}
+
+// IsL3Only returns false if the given L4RuleContexts contains any entry. If it
+// does not contain any entry it is considered a L3 only rule.
+func (rc L4RuleContexts) IsL3Only() bool {
+	return rc != nil && len(rc) == 0
+}
+
+// L4RuleContext represents a L4 rule
 // Don't use pointers here since this structure is used as key on maps.
-type RuleContext struct {
-	// SecID is the security ID for the numeric identity
-	SecID NumericIdentity
+type L4RuleContext struct {
 	// Port is the destination port in the policy in network byte order
 	Port uint16
 	// Proto is the protocol ID used
 	Proto uint8
-	// L7RedirectPort is the L7 redirect port in the policy in network byte order
-	L7RedirectPort uint16
-	// IsRedirect is set in case the rule is a redirect
-	IsRedirect bool
+}
+
+type L7RuleContext struct {
+	// RedirectPort is the L7 redirect port in the policy in network byte order
+	RedirectPort uint16
+	// L4Installed specifies if the given rule is installed in the L4 BPF map
+	L4Installed bool
```

Comment:
`InstalledInL4` or `L4Dependent` might have been more clear when I first saw this variable in other places. I'm not sure that my suggestions are actually better for anyone else.

Reply:
Done

Reply:
Thanks :) Maybe someone can make an AI for naming variables... instead of silly things like self-driving cars.